providers/oidc.go

```diff
@@ -52,23 +42,74 @@ func (p *OIDCProvider) Redeem(ctx context.Context, redirectURL, code string) (s
 	}
 	token, err := c.Exchange(ctx, code)
 	if err != nil {
-		return nil, fmt.Errorf("token exchange: %v", err)
+		return nil, fmt.Errorf("token exchange failure: %v", err)
 	}
 
-	// in the initial exchange the id token is mandatory
-	idToken, err := p.findVerifiedIDToken(ctx, token)
-	if err != nil {
-		return nil, fmt.Errorf("could not verify id_token: %v", err)
-	} else if idToken == nil {
-		return nil, fmt.Errorf("token response did not contain an id_token")
+	return p.createSession(ctx, token, false)
+}
+
+// EnrichSessionState is called after Redeem to allow providers to enrich session fields
+// such as User, Email, Groups with provider specific API calls.
+func (p *OIDCProvider) EnrichSession(ctx context.Context, s *sessions.SessionState) error {
+	if p.ProfileURL.String() == "" {
+		if s.Email == "" {
+			return errors.New("id_token did not contain an email and profileURL is not defined")
+		}
+		return nil
 	}
 
-	s, err = p.createSessionState(ctx, token, idToken)
+	// Try to get missing emails or groups from a profileURL
+	if s.Email == "" || len(s.Groups) == 0 {
+		err := p.callProfileURL(ctx, s)
+		if err != nil {
+			logger.Errorf("Warning: Profile URL request failed: %v", err)
+		}
+	}
+
+	// If a mandatory email wasn't set, error at this point.
+	if s.Email == "" {
+		return errors.New("neither the id_token nor the profileURL set an email")
+	}
+	return nil
+}
+
+// callProfileURL enriches a session's Email & Groups via the JSON response of
+// an OIDC profile URL
+func (p *OIDCProvider) callProfileURL(ctx context.Context, s *sessions.SessionState) error {
+	respJSON, err := requests.New(p.ProfileURL.String()).
+		WithContext(ctx).
+		WithHeaders(makeOIDCHeader(s.AccessToken)).
+		Do().
+		UnmarshalJSON()
 	if err != nil {
-		return nil, fmt.Errorf("unable to update session: %v", err)
+		return err
+	}
+
+	email, err := respJSON.Get(p.EmailClaim).String()
+	if err == nil && s.Email == "" {
+		s.Email = email
+	}
+
+	// Handle array & singleton groups cases
```

Comment:
This feels ugly... Have a better idea for singtleton + list support from JSON methods?

Reply:
Someone has to coerce it to a list, if the json doesn't do that automagically we have to. Might be nice as a helper though?

Reply:
Yeah, I'll whip up a helper for this. That feels cleaner and will reduce the conditional complexity of this method.